### PR TITLE
[String] Detangling _StringGuts; clean up UTF8View some

### PR DESCRIPTION
**Pull request by milseman**
Trying to de-tangle _StringGuts from the assumption that it's a
collection of UTF16 code units. In the process, clean up
String.UTF8View, outline cold paths, small string fast paths, etc.

### CODE REVIEW

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
@swift-ci please smoke test compiler performance

**Comment by milseman** [comment]
@eeckstein a side effect here is to further reduce code size with outlining slow paths and improve perf with small string fast-paths.

**Comment by milseman** [comment]
(still WIP though, I have lots more to do along those lines as well)

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
@swift-ci please smoke test compiler performance

**Comment by shahmishal** [comment]
We are testing new smoke benchmark without Onone 

@swift-ci smoke benchmark

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
@swift-ci please smoke test compiler performance

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
@swift-ci please smoke test compiler performance

**Comment by shahmishal** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
Most of the append slowdown was because of an accidental closure capture of self. There's still around a 40% regression left, though.

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
Well, that's showing improvements to ArrayAppendUTF16Substring while locally I still show a regression.

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
@swift-ci please smoke test complier performance

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
@swift-ci please smoke test compiler performance

File: stdlib/public/core/StringUTF8.swift [hunk 1]
```diff
@@ -545,33 +614,51 @@ extension String.UTF8View.Iterator : IteratorProtocol {
...
+  case 0x80..<0x800: return 2
+  case 0x800..<0xDC00: return 3
+  case 0xDC00..<0xE000: return UTF16.isLeadSurrogate(prev) ? 1 : 3
+  default: return 3
+  }
+}
+
 extension String.UTF8View {
-  @inlinable // FIXME(sil-serialize-all)
+  @usableFromInline
```
Comment (lorentey) [inline stdlib/public/core/StringUTF8.swift]: I think this is not used from inlinable code anymore, so `@usableFromInline` may be removed.

**Review by lorentey — APPROVED** [review]
This looks good! 👍 Great to see the iterator gone.

I assume the plan is to rename `_isASCIIOrSmallASCII` to `isASCII` later, right?

I suspect some of these `@effects(releasenone)`s can be upgraded to `@effects(readonly)`. (Although an opaque NSString subclass may violate both, I guess.)

**Comment by milseman** [comment]
@eeckstein 1.4-1.5x speedup for `Array(mySubstring.utf8)` when mySubstring is not ASCII, and ~3x when it is ;-)

File: stdlib/public/core/StringUTF8.swift (hunk 1 above)
Comment (milseman) [reply stdlib/public/core/StringUTF8.swift]: Ah, good point. I had done a purge of many of these in another PR and I'll try another one. Also, I think in general we could use a diagnostic pass letting us know.

https://bugs.swift.org/browse/SR-7687

**Pull request merged**